examples/recurrent_bidirectional.py

```diff
@@ -0,0 +1,144 @@
+import numpy as np
+import theano
+import theano.tensor as T
+import lasagne
+# Sequence length
+LENGTH = 10
+# Number of units in the hidden (recurrent) layer
+N_HIDDEN = 4
+# Number of training sequences in each batch
+N_BATCH = 30
+# Delay used to generate artificial training data
+DELAY = 2
+# SGD learning rate
+LEARNING_RATE = 1e-5
+# Number of iterations to train the net
+N_ITERATIONS = 1000
+
+
+def gen_data(length=LENGTH, n_batch=N_BATCH, delay=DELAY):
+    '''
+    Generate a simple lag sequence
+
+    :parameters:
+        - length : int
+            Length of sequences to generate
+        - n_batch : int
+            Number of training sequences per batch
+        - delay : int
+            How much to delay one feature dimension in the target
+
+    :returns:
+        - X : np.ndarray, shape=(n_batch, length, 2)
+            Input sequence
+        - y : np.ndarray, shape=(n_batch, length, 1)
+            Target sequence, where
+            y[n] = X[:, n, 0] - X[:, n - delay, 1] + noise
+    '''
+    X = np.random.rand(n_batch, length, 2)
+    y = X[:, :, 0].reshape((n_batch, length, 1))
+    # Compute y[n] = X[:, n, 0] - X[:, n - delay, 1] + noise
+    y[:, delay:, 0] -= (X[:, :-delay, 1]
+                        + .01*np.random.randn(n_batch, length - delay))
+    return X.astype(theano.config.floatX), y.astype(theano.config.floatX)
+
+
+# Generate a "validation" sequence whose cost we will periodically compute
+X_val, y_val = gen_data()
+mask_val = np.ones(shape=(N_BATCH, LENGTH), dtype=theano.config.floatX)
+
+# Construct vanilla RNN: One recurrent layer (with input weights) and one
+# dense output layer
+l_in = lasagne.layers.InputLayer(shape=(N_BATCH, LENGTH, X_val.shape[-1]))
+
+# This input layer is used to tell the input-to-hidden what shape to expect
+# As we iterate over time steps, the input will be batch size x feature dim
+l_recurrent_in_fwd = lasagne.layers.InputLayer(shape=(N_BATCH, X_val.shape[-1]))
+l_input_to_hidden_fwd = lasagne.layers.DenseLayer(l_recurrent_in_fwd, N_HIDDEN,
+                                              nonlinearity=None)
```

Comment:
Just in case you didn't notice: right now this comment is a bit hidden behind this "show outdated diff" thing.  Maybe you want to elevate this to become a proper comment on the pull request.

Your argument regarding point number 3 seems good.  But sadly I've had too little exposure to this code and to LSTMs to be of any real help with the other issues you're asking about.